enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,545 @@
+infrastructure services.
+
+These capabilities include:
+- Highly available and dynamically routed load-balanced api access
+- Highly available and dynamically routed ingress access
+
+The assets needed for these capabilities implementation are rendered by the [MCO](https://github.com/openshift/machine-config-operator) and
+the [baremetal-runtimecfg](https://github.com/openshift/baremetal-runtimecfg) is used significantly within the manifests and templates of
+the machine-config-operator project.
+There are also changes in the API and installer projects, which are further described in sections below.
+
+### Workflow Description
+
+The installer accepts an LB type (default remains what it does today), which will be set into the Cluster Infrastructure Object. Then MCO will read the LB type and if it's BGP it'll not deploy Keepalived/HAproxy. Instead, it'll deploy FRR (a dynamic routing protocol suite, explained in details later) and configure it to peer with the BGP infrastructure.
+
+### Spine-Leaf architecture
+
+![image](on-prem-bgp-networking-spine-leaf.png)
+
+This architecture is a common design for data centers that have a spine-leaf architecture.
+The OpenShift nodes are connected to the leafs, which are connected to the spine.
+In our proposal, the control plane will peer with the leafs, and the leafs will peer with the spine, so the VIPs
+will be routable across the network.
+
+### API
+
+In conjunction to the work that has been (and is being) done on Failure Domains, we will propose 
+a new [API](https://github.com/openshift/api/pull/1321) to configure Load Balancers, so we will be able to:
+- Change the load-balancer type (default remains VRRP), to be able to enable BGP.
+- If BGP is enabled, allow to provide a list of BGP speakers and their configurations.
+- The Machine Config Operator (MCO) will monitor this setting and generate the required configuration.
+
+There will be some change in the API:
+
+```golang
+type OpenStackAPIBGPPeer struct {
+	// asn is the Autonomous System number of the peer.
+	//
+	// +optional
+	ASN string `json:"asn,omitempty"`
+
+	// ip is the IP address of the peer, as reachable from the Control
+	// plane machine. It may be either IPv4 or IPv6
+	//
+	// +kubebuilder:validation:Required
+	// + ---
+	// + {IPv4 or IPv6} validation in crd.yaml-patch
+	IP string `json:"ip"`
+
+	// password for BGP authentication against the peer
+	//
+	// +optional
+	Password string `json:"password,omitempty"`
+}
+
+// OpenStackAPIBGPSpeaker describes the BGP autonomous system that will contain
+// the API VIP for a specific failure domain.
+type OpenStackAPIBGPSpeaker struct {
+	// failureDomain is the name of a failure domain which this BGP
+	// configuration applies to. A failure domain with that name must be
+	// defined in the OpenStack platform spec. If there are no failure
+	// domains defined, use "default".
+	//
+	// +kubebuilder:validation:Required
+	FailureDomain string `json:"failureDomain"`
```

Comment:
I propose this to be "Subnet" so that the BGP configuration is independent from the Failure Domains config (see [this change](https://github.com/openshift/api/commit/d985194c0c6b96c07a07e75133d17eb0a4dcb86f#diff-07b264a49084976b670fb699badaca1795027d6ea732a99226a5388104f6174f))

Reply:
I removed all API code for now, since we're discuss this in the openshift/api PR. I'll put it back here once we decide what's best. BTW nice proposal on using the subnet instead of failure domain name. I like it.